src/PureScript/Backend/Optimizer/Semantics.purs

```diff
@@ -603,105 +658,134 @@ makeLet ident binding go = case binding of
     SemLet ident binding go
 
 deref :: BackendSemantics -> BackendSemantics
-deref = case _ of
-  NeutLocal _ _ (Just expr) -> expr
-  expr -> expr
+deref = fromMaybe <*> go
+  where
+  go = case _ of
+    NeutAccessor expr' accessand -> go expr' >>=
+      case _, accessand of
+        NeutLit (LitArray arr), GetIndex i
+          | Just v <- Array.index arr i -> go v
+        NeutLit (LitRecord rec), GetProp p
+          | Just (Prop _ v) <- Array.find (\(Prop k _) -> k == p) rec -> go v
+        NeutData _ _ _ _ vals, GetCtorField _ _ _ _ _ i
+          | Just (Tuple _ v) <- Array.index vals i -> go v
+        _, _ -> Nothing
+    NeutLocal _ _ expr -> expr >>= go
+    e@(NeutLit (LitInt _)) -> Just e
+    e@(NeutLit (LitNumber _)) -> Just e
+    e@(NeutLit (LitString _)) -> Just e
+    e@(NeutLit (LitChar _)) -> Just e
+    e@(NeutLit (LitBoolean _)) -> Just e
+    -- in the following three cases, all we may need
+    -- is the object itself, meaning that we don't want to
+    -- give up on the whole thing if it doesn't work
+    -- so we call `deref` on the children
+    -- instead of calling `go`
+    NeutLit (LitArray arr) -> Just (NeutLit (LitArray (deref <$> arr)))
+    NeutLit (LitRecord rec) -> Just (NeutLit (LitRecord ((map <<< map) deref rec)))
+    NeutData qual ct ty tag vals -> Just (NeutData qual ct ty tag ((map <<< map) deref vals))
+    -- fail
+    _ -> Nothing
```

Comment:
Why is all of this here? Whether or not this should be included is a completely separate issue, and I don't think it should be included in this PR. There may be a completely different way to accomplish this, and I would prefer not to consider this in order to consider more let floating.

Reply:
In trying to come up with tests that replicates the issue I had in my larger branch, this was the minimum amount of code I had to copy over from that branch to make the tests work. I'm all for brainstorming different ways to accomplish this, however (I'll address that in my next comment).